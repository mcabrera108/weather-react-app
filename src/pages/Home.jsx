import style from "../style_modules/layout.module.css";
import { useMediaQuery } from "react-responsive";
import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMagnifyingGlass } from "@fortawesome/free-solid-svg-icons";
import DataModule from "../components/ui/DataModule";

function Home() {
  const [weather, setWeather] = useState(null);
  const [isFahrenheit, setIsFahrenheit] = useState(true);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);
  const WEATHER_KEY = import.meta.env.VITE_WEATHER_KEY;

  const isMobileDevice = useMediaQuery({
    query: `(max-width:1024px)`,
  });

  const [weatherInput, setWeatherInput] = useState("Seattle");

  async function getWeather(event) {
    event.preventDefault();
    try {
      const response = await fetch(
        `https://api.weatherapi.com/v1/current.json?key=${WEATHER_KEY}&q=${weatherInput}&days=3`
      );
      const actualData = await response.json();

      console.log(actualData);
      if (!response.ok) {
        throw new Error(
          `This is an HTTP Error: The status is ${response.data}`
        );
      }
      setWeather(actualData);
      setError(null);
    } catch (err) {
      setError(err.message);
      setWeather(null);
    } finally {
      setLoading(false);
    }
  }

  return (
    <div className={style.mainContainer}>
      <div className={style.homeContainer}>
        <div className={style.homeWrapper}>
          <div className={style.searchContainer}>
            <form onSubmit={getWeather}>
              <input
                type="text"
                onChange={(e) => {
                  setWeatherInput(e.target.value);
                }}
                size={isMobileDevice ? 20 : 50}
                className={style.userInput}
                placeholder="Search for a location..."
              />
              <button
                type="submit"
                disabled={loading}
                className={style.userInputBtn}
              >
                {loading ? (
                  "Loading..."
                ) : (
                  <FontAwesomeIcon
                    icon={faMagnifyingGlass}
                    size="lg"
                    color="white"
                  />
                )}
              </button>
            </form>
          </div>
          <div className={style.areaContainer}>
            {error && <p>Unable to fetch details for this submission</p>}
            {weather && (
              <p>
                {weather.location.name}
                {weather.location.region ? `, ${weather.location.region}` : ""}
              </p>
            )}
          </div>
          <div className={style.tempContainer}>
            {weather && (
              <p
                onClick={() => {
                  setIsFahrenheit(!isFahrenheit);
                }}
              >
                {isFahrenheit ? (
                  <span>{weather.current.temp_f} &deg;F</span>
                ) : (
                  <span>{weather.current.temp_c} &deg;C</span>
                )}
              </p>
            )}
          </div>
          <div className={style.feelsLikeContainer}>
            {weather && (
              <span>Click the temp above to toggle from &deg;F to &deg;C</span>
            )}
          </div>
          <div className={style.weatherContainer}>
            {weather && (
              <img
                src={weather.current.condition.icon}
                className={style.weatherIcon}
              />
            )}
            {weather && <p>{weather.current.condition.text}</p>}
          </div>
          <div className={style.feelsLikeContainer}>
            {/*weather && <img src={weather.current.condition.icon} />*/}
            {weather && (
              <p>
                Feels Like:{" "}
                {isFahrenheit
                  ? weather.current.feelslike_f
                  : weather.current.feelslike_c}
              </p>
            )}
          </div>
          <div className={style.allModContainer}>
            {weather && (
              <DataModule weather={weather} isFahrenheit={isFahrenheit} />
            )}
          </div>
        </div>
      </div>
    </div>
  );
}
export default Home;
