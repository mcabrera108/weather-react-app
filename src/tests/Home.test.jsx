import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import Home from "../pages/Home";

describe("Home Component", () => {
  it("Renders Home without any api calls made", () => {
    const container = render(<Home />);
    expect(container).toMatchSnapshot();
  });
  it("Renders input search bar when Home Component is rendered", () => {
    const container = render(<Home />);
    expect(container.getByPlaceholderText(/Search/)).toHaveAttribute(
      "placeholder",
      "Search for a location..."
    );
  });
  it("Renders correct icon from the React Icons Library", () => {
    const container = render(<Home />);
    const svgSelector = container.getByTitle("magnifyingglass");
    expect(svgSelector).toBeInTheDocument();
  });
});
