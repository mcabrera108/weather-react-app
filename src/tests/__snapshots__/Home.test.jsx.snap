// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Component > Renders Home without any api calls made 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_mainContainer_4f5d64"
      >
        <div
          class="_homeContainer_4f5d64"
        >
          <div
            class="_homeWrapper_4f5d64"
          >
            <div
              class="_searchContainer_4f5d64"
            >
              <form>
                <input
                  class="_userInput_4f5d64"
                  placeholder="Search for a location..."
                  size="50"
                  type="text"
                />
                <button
                  class="_userInputBtn_4f5d64"
                  type="submit"
                >
                  <svg
                    aria-labelledby="svg-inline--fa-title-yAc12awxQGcv"
                    class="svg-inline--fa fa-magnifying-glass fa-xl "
                    color="white"
                    data-icon="magnifying-glass"
                    data-prefix="fas"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title
                      id="svg-inline--fa-title-yAc12awxQGcv"
                    >
                      magnifyingglass
                    </title>
                    <path
                      d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </form>
            </div>
            <div
              class="_areaContainer_4f5d64"
            />
            <div
              class="_tempContainer_4f5d64"
            />
            <div
              class="_feelsLikeContainer_4f5d64"
            />
            <div
              class="_weatherContainer_4f5d64"
            />
            <div
              class="_feelsLikeContainer_4f5d64"
            />
            <div
              class="_allModContainer_4f5d64"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_mainContainer_4f5d64"
    >
      <div
        class="_homeContainer_4f5d64"
      >
        <div
          class="_homeWrapper_4f5d64"
        >
          <div
            class="_searchContainer_4f5d64"
          >
            <form>
              <input
                class="_userInput_4f5d64"
                placeholder="Search for a location..."
                size="50"
                type="text"
              />
              <button
                class="_userInputBtn_4f5d64"
                type="submit"
              >
                <svg
                  aria-labelledby="svg-inline--fa-title-yAc12awxQGcv"
                  class="svg-inline--fa fa-magnifying-glass fa-xl "
                  color="white"
                  data-icon="magnifying-glass"
                  data-prefix="fas"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title
                    id="svg-inline--fa-title-yAc12awxQGcv"
                  >
                    magnifyingglass
                  </title>
                  <path
                    d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </form>
          </div>
          <div
            class="_areaContainer_4f5d64"
          />
          <div
            class="_tempContainer_4f5d64"
          />
          <div
            class="_feelsLikeContainer_4f5d64"
          />
          <div
            class="_weatherContainer_4f5d64"
          />
          <div
            class="_feelsLikeContainer_4f5d64"
          />
          <div
            class="_allModContainer_4f5d64"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
